### PR TITLE
[SPARK-16484][SQL] Add HyperLogLogPlusPlus sketch generators/evaluator/aggregators

### PR DESCRIPTION
**Pull request by RyanBerti**
### What changes were proposed in this pull request?
This PR converts the existing HyperLogLogPlusPlus case class into a trait, and adds a set of case classes which extend that trait and provide new functionality to generate/evaluate/aggregate the serialized version of the HLL++ sketch. 

### Why are the changes needed?
These additional functions are being added to support HLL++ re-aggregation, which can provide significant performance improvements when running distinct counts over time series datasets. 

### Does this PR introduce _any_ user-facing change?
Yes, this PR adds two new aggregate functions, approx_count_distinct_sketch and approx_count_distinct_agg_sketch. This PR also introduces a non-aggregate function approx_count_distinct_eval_sketch.

### How was this patch tested?
Unit tests were added to the HyperLogLogPlusPlusSuite and the DataFrameAggregateSuite to test sketch generation, evaluation, and aggregation.

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/FunctionRegistry.scala [hunk 1]
```diff
@@ -455,6 +455,9 @@ object FunctionRegistry {
 
     // aggregate functions
     expression[HyperLogLogPlusPlus]("approx_count_distinct"),
+    expression[HyperLogLogPlusPlusSketch]("approx_count_distinct_sketch"),
```
Comment (dtenedor) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/FunctionRegistry.scala:458]: I would recommend to give these shorter names for easier usability in SQL. For example, in prior systems, I've used: HLL_COUNT.INIT, HLL_COUNT.MERGE, and HLLCOUNT.FINALIZE. What do you think?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/HyperLogLogPlusPlusAggSketch.scala [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+    Examples:
+      > SELECT _FUNC_(approx_count_distinct_sketch(col1))
+        FROM VALUES (1), (1), (2), (2), (3) tab(col1);
+       3
+  """,
+  group = "agg_funcs",
+  since = "3.3.1")
+case class HyperLogLogPlusPlusAggSketch(
+    child: Expression,
+    var relativeSD: Double = 0.05,
```
Comment (dtenedor) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/HyperLogLogPlusPlusAggSketch.scala]: can we comment about these arguments in the class comment?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/HyperLogLogPlusPlusAggSketch.scala [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+    mutableAggBufferOffset: Int = 0,
+    inputAggBufferOffset: Int = 0) extends HyperLogLogPlusPlusTrait {
+
+  def this(child: Expression) = {
+    this(child = child, relativeSD = 0.05, mutableAggBufferOffset = 0, inputAggBufferOffset = 0)
```
Comment (dtenedor) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/HyperLogLogPlusPlusAggSketch.scala]: here we repeat the magic number value, can we dedup into one place?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/HyperLogLogPlusPlusAggSketch.scala [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+    val v = child.eval(input)
+    if (v != null) {
+
+      val (newRelativeSD, newInternalRow) =
+        HyperLogLogPlusPlusSketch.deserializeSketch(v.asInstanceOf[Array[Byte]])
+
+      relativeSDUpdated match {
+        case true =>
+          if (newRelativeSD != relativeSD) {
+            throw new IllegalStateException("One or more of the HyperLogLogPlusPlusSketch " +
```
Comment (dtenedor) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/HyperLogLogPlusPlusAggSketch.scala]: should we put the values in the error message to help with debugging? Can we think of any other information that we could add in here to help? If this is illegal state, we don't have to worry about the user-friendliness so much and can just add debug information...

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/HyperLogLogPlusPlusAggSketch.scala [hunk 5]
```diff
@@ -0,0 +1,183 @@
...
+ *
+ * HyperLogLog in Practice: Algorithmic Engineering of a State of The Art Cardinality Estimation
+ * Algorithm
+ * http://static.googleusercontent.com/external_content/untrusted_dlcp/research.google.com/en/us/pubs/archive/40671.pdf
+ *
+ * Appendix to HyperLogLog in Practice: Algorithmic Engineering of a State of the Art Cardinality
+ * Estimation Algorithm
+ * https://docs.google.com/document/d/1gyjfMHy43U9OWBXxfaeG-3MjGzejW1dlpyMwEYAAWEI/view?fullscreen#
+ *
+ * @param child the HLL++ sketch to be aggregated and evaluated.
```
Comment (dtenedor) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/HyperLogLogPlusPlusAggSketch.scala]: I think it could be helpful to have a brief diagram of what the buffer looks like, maybe some ASCII block showing what the headers look like, etc. Could we show an example and illustrate how the buffer is updated?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/HyperLogLogPlusPlusAggSketch.scala [hunk 6]
```diff
@@ -0,0 +1,183 @@
...
+   * HyperLogLogPlusPlusSketch instance, which utilized a HLL++ helper that may have been
+   * configured differently than the one provided by our super class. Because we can't
+   * re-instantiate the super class's HLL++ helper, let's maintain the first instance that
+   * we deserialize, and ensure that our instance is able to fit within the words
+   * allocated in the aggregation buffer by the super class's HLL++ helper instance.
+   */
+  private var relativeSDUpdated: Boolean = false
+
+  /**
+   * Update the HLL++ buffer.
```
Comment (dtenedor) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/HyperLogLogPlusPlusAggSketch.scala]: mention that this consumes an input row?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/HyperLogLogPlusPlusAggSketch.scala [hunk 7]
```diff
@@ -0,0 +1,183 @@
...
+
+          // Then fill the rest of the agg buffer with the words from the sketch
+          (0 until hllppHelper.numWords).foreach(i => {
+            buffer.setLong(mutableAggBufferOffset + 1 + i, newInternalRow.getLong(i))
+          })
+      }
+    }
+  }
+
+  override def merge(buffer1: InternalRow, buffer2: InternalRow): Unit = {
```
Comment (dtenedor) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/HyperLogLogPlusPlusAggSketch.scala:173]: please add a comment for this, mentioning that it consumes a previous sketch?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/HyperLogLogPlusPlusAggSketch.scala [hunk 8]
```diff
@@ -0,0 +1,183 @@
...
+    val numWords1 = buffer1.getLong(mutableAggBufferOffset)
+    val numWords2 = buffer2.getLong(inputAggBufferOffset)
+    if (numWords1 != 0 && numWords2 != 0 && numWords1 != numWords2) {
+      throw new IllegalStateException("One or more of the HyperLogLogPlusPlusSketch " +
```
Comment (dtenedor) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/HyperLogLogPlusPlusAggSketch.scala]: same, should we add more debug info here?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/HyperLogLogPlusPlusAggSketch.scala [hunk 9]
```diff
@@ -0,0 +1,183 @@
...
+        "sketches were configured with different relativeSD values, " +
+        "and cannot be aggregated")
+    }
+
+    // we need to explicitly set numWords so that the buffer retains it between merges
```
Comment (dtenedor) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/HyperLogLogPlusPlusAggSketch.scala]:
```suggestion
    // We need to explicitly set numWords so that the buffer retains it between merges
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/HyperLogLogPlusPlusAggSketch.scala [hunk 10]
```diff
@@ -0,0 +1,183 @@
...
+        case true =>
+          if (newRelativeSD != relativeSD) {
+            throw new IllegalStateException("One or more of the HyperLogLogPlusPlusSketch " +
+              "sketches were configured with different relativeSD values, " +
+              "and cannot be aggregated")
+          }
+
+          // When merging into our agg buffer, skip the first long; the
+          // deserialized row is formatted without the leading long so no
+          // skipping is necessary
```
Comment (dtenedor) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/HyperLogLogPlusPlusAggSketch.scala]: can we add a period at the end of this comment, here and elsewhere in this PR?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/HyperLogLogPlusPlusAggSketch.scala [hunk 11]
```diff
@@ -0,0 +1,183 @@
...
+    if (numWords1 != 0 && numWords2 != 0 && numWords1 != numWords2) {
+      throw new IllegalStateException("One or more of the HyperLogLogPlusPlusSketch " +
+        "sketches were configured with different relativeSD values, " +
+        "and cannot be aggregated")
+    }
+
+    // we need to explicitly set numWords so that the buffer retains it between merges
+    buffer1.setLong(mutableAggBufferOffset, if (numWords1 == 0) numWords2 else numWords1)
+
+    // When merging agg buffers, skip the first long of both
```
Comment (dtenedor) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/HyperLogLogPlusPlusAggSketch.scala]: please mention why?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/HyperLogLogPlusPlusAggSketch.scala [hunk 12]
```diff
@@ -0,0 +1,183 @@
...
+    hllppHelper.merge(buffer1 = buffer1, buffer2 = buffer2,
+      offset1 = mutableAggBufferOffset + 1, offset2 = inputAggBufferOffset + 1)
+  }
+
+  override def eval(buffer: InternalRow): Any = {
+    // When evaling agg buffer, skip the first long
```
Comment (dtenedor) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/HyperLogLogPlusPlusAggSketch.scala]: same?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/HyperLogLogPlusPlusSketch.scala [hunk 13]
```diff
@@ -0,0 +1,143 @@
...
+ * Algorithm
+ * http://static.googleusercontent.com/external_content/untrusted_dlcp/research.google.com/en/us/pubs/archive/40671.pdf
+ *
+ * Appendix to HyperLogLog in Practice: Algorithmic Engineering of a State of the Art Cardinality
+ * Estimation Algorithm
+ * https://docs.google.com/document/d/1gyjfMHy43U9OWBXxfaeG-3MjGzejW1dlpyMwEYAAWEI/view?fullscreen#
+ *
+ * @param child to estimate the cardinality of.
+ * @param relativeSD the maximum relative standard deviation allowed.
+ */
```
Comment (dtenedor) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/HyperLogLogPlusPlusSketch.scala:51]: same, let's add an example and diagram to show what the protocol of the buffer is?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/HyperLogLogPlusPlusSketch.scala [hunk 14]
```diff
@@ -0,0 +1,143 @@
...
+  examples = """
+    Examples:
+      > SELECT approx_count_distinct(_FUNC_(col1)) FROM VALUES (1), (1), (2), (2), (3) tab(col1);
+       3
+  """,
+  group = "agg_funcs",
+  since = "3.3.1")
+case class HyperLogLogPlusPlusSketch(
+    child: Expression,
+    relativeSD: Double = 0.05,
```
Comment (dtenedor) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/HyperLogLogPlusPlusSketch.scala]: same, can you please comment these arguments?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/HyperLogLogPlusPlusSketch.scala [hunk 15]
```diff
@@ -0,0 +1,143 @@
...
+          }
+        }
+
+        (relativeSD, row)
+      case _ => throw new UnsupportedOperationException()
+    }
+
+
+  }
+}
```
Comment (dtenedor) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/HyperLogLogPlusPlusSketch.scala]: please add a newline at the end of the file

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/HyperLogLogPlusPlusSketch.scala [hunk 16]
```diff
@@ -0,0 +1,143 @@
...
+      byteBuffer.putLong(buffer.getLong(mutableAggBufferOffset + i))
+    }
+
+    byteBuffer.array()
+  }
+
+  def deserializeSketch(sketch: Array[Byte]):
+    (Double, InternalRow) = {
+
+    val byteBuffer = ByteBuffer.wrap(sketch)
```
Comment (dtenedor) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/HyperLogLogPlusPlusSketch.scala:148]: this implementation block is a bit tricky. Can you please add a comment?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/misc.scala [hunk 17]
```diff
@@ -432,4 +433,35 @@ case class AesDecrypt(
     copy(newChildren(0), newChildren(1), newChildren(2), newChildren(3))
   }
 }
+
+@ExpressionDescription(
+  usage = """_FUNC_(expr) - Return the estimated distinct count associated with this HyperLogLog++ sketch.""",
```
Comment (dtenedor) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/misc.scala:438]: same, let's please add an example and diagram to show the protocol of the sketch buffer in here?

**Review by dtenedor — COMMENTED** [review]
Thanks @RyanBerti for putting in the work to implement this! Did a review round and left some comments.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/FunctionRegistry.scala (hunk 1 above)
Comment (RyanBerti) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/FunctionRegistry.scala:458]: My only worry with this is that by diverging from the 'HyperLogLogPlusPlus' and 'approx_count_distinct' prefixes, the new functions wouldn't be visible in alphabetical order alongside the original. I would prefer the function names to be shorter, but I felt that using the same prefix was a higher priority. Would you mind if we have a larger discussion about naming in the main PR and get opinions from a wider group?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/HyperLogLogPlusPlusAggSketch.scala (hunk 2 above)
Comment (RyanBerti) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/HyperLogLogPlusPlusAggSketch.scala]: Good catch, will do.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/HyperLogLogPlusPlusAggSketch.scala (hunk 3 above)
Comment (RyanBerti) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/HyperLogLogPlusPlusAggSketch.scala]: Yep, will add default val to the HyperLogLogPlusPlus object and reference it from others.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/HyperLogLogPlusPlusAggSketch.scala (hunk 4 above)
Comment (RyanBerti) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/HyperLogLogPlusPlusAggSketch.scala]: Good call, will do.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/HyperLogLogPlusPlusSketch.scala (hunk 16 above)
Comment (RyanBerti) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/HyperLogLogPlusPlusSketch.scala:148]: Added comments, let me know if they're helpful?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/HyperLogLogPlusPlusSketch.scala (hunk 14 above)
Comment (RyanBerti) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/HyperLogLogPlusPlusSketch.scala]: Child / relativeSD are commented, but not buffer offset args - the original HyperLogLogPlusPlus didn't have those commented as I don't think they're supposed to be exposed to the end user.

**Comment by RyanBerti** [comment]
Hi @dtenedor thanks for your review! I've resolved all of the comments where I made changes; if you wouldn't mind re-reviewing that'd be appreciated. Once you re-review I'll open the updated PR against the main Spark repo.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/FunctionRegistry.scala (hunk 1 above)
Comment (dtenedor) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/FunctionRegistry.scala:458]: Yeah that's fine. We can also implement it this way and add aliases later if we want.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/HyperLogLogPlusPlusAggSketch.scala [hunk 18]
```diff
@@ -0,0 +1,202 @@
...
+  override def update(buffer: InternalRow, input: InternalRow): Unit = {
+    val v = child.eval(input)
+    if (v != null) {
+
+      val (newRelativeSD, newInternalRow) =
+        HyperLogLogPlusPlusSketch.deserializeSketch(v.asInstanceOf[Array[Byte]])
+
+      relativeSDUpdated match {
+        case true =>
+          if (newRelativeSD != relativeSD) {
```
Comment (dtenedor) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/HyperLogLogPlusPlusAggSketch.scala]: here we are comparing floating-point numbers for exact inequality. Are we certain there are no false positives/negatives for this? Same for the `if (newRelativeSD < relativeSD)` below?

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/aggregate/HyperLogLogPlusPlusSuite.scala [hunk 19]
```diff
@@ -175,4 +195,45 @@ class HyperLogLogPlusPlusSuite extends SparkFunSuite {
...
+  test("SPARK-16484: Test sketch evaluation") {
+    val (hll, input, buffer) = createSketchEstimator(0.05, DoubleType)
+    List(1, 1, 2, 2, 3).foreach(i => {
+      input.setDouble(0, i)
+      hll.update(buffer, input)
+    })
+    val sketch = hll.eval(buffer).asInstanceOf[Array[Byte]]
+    val bb = ByteBuffer.wrap(sketch)
+    assert(bb.getInt() == 1)
+    assert(bb.getDouble() == 0.05)
```
Comment (dtenedor) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/aggregate/HyperLogLogPlusPlusSuite.scala:313]: same, this performs floating-point comparison; should we use an epsilon for determinism in the testing?

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/aggregate/HyperLogLogPlusPlusSuite.scala [hunk 20]
```diff
@@ -175,4 +195,45 @@ class HyperLogLogPlusPlusSuite extends SparkFunSuite {
...
+
+    // unclear if this is correctly reproducing round trip spark behavior via EmptyRow input
+    val approxDistinctCount = HyperLogLogPlusPlusEvalSketch(Literal(sketch)).eval(EmptyRow)
+    assert(approxDistinctCount == 3)
```
Comment (dtenedor) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/aggregate/HyperLogLogPlusPlusSuite.scala:317]: we will need more test coverage for this feature. Can you please add corner cases including all supported input data types, NULL values, min and max values of each supported data type, any overflow conditions, and all error cases?

**Review by dtenedor — COMMENTED** [review]
Thanks a lot for adding the new comments, they are very helpful for understanding the format! This is probably ready for wider review now, we just need to add some more test cases.

**Comment by dtenedor** [comment]
So far in this proposal, we're converting the relativeSD into the logm value like this:

```
val p = Math.ceil(2.0d * Math.log(1.106d / relativeSD) / Math.log(2.0d)).toInt
```

It would be preferable to store the actual logm value instead of the relativeSD. We rarely need to use the relativeSD so we'll always have to convert to the logm, which begs the question why don't we just store the logm (not to mention that it's smaller and more intuitive since relativeSD's of 0.05, 0.06, etc. all come down to a logm of 9, so we don't exactly guarantee that it's not going to be slightly different from what the user specified.)

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/HyperLogLogPlusPlusAggSketch.scala (hunk 18 above)
Comment (RyanBerti) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/HyperLogLogPlusPlusAggSketch.scala]: I modified this to test the Integer numWords values derived from the Double relativeSD values so we don't have to implement some custom precision comparison functions.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/aggregate/HyperLogLogPlusPlusSuite.scala (hunk 19 above)
Comment (RyanBerti) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/aggregate/HyperLogLogPlusPlusSuite.scala:313]: I left this unmodified as I think comparisons of that 0.05 value are deterministic, and I don't expect the default relativeSD value to change.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/aggregate/HyperLogLogPlusPlusSuite.scala (hunk 20 above)
Comment (RyanBerti) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/aggregate/HyperLogLogPlusPlusSuite.scala:317]: I've extended the test suite to test the sketch generator/aggregators using the test harness that's used for testing the existing HLL++ functionality. I've got the sketch aggregation error cases tested via the negative tests added to the DataframeAggregationSuite. I'm open to adding more tests, though I could definitely use some guidance on what tests would be the most valuable given what's already already covered (IE I could probably build out some more tests that generate random data for each datatype, but I think that ends up testing the core HLL++ functionality more than the new sketch generators/aggregators as the generators/aggregators always operate on BinaryType data).

**Comment by RyanBerti** [comment]
@dtenedor regarding storing the P value, instead of relativeSD with the serialized sketch: the P value is currently private within the HyperLogLogPlusPlusHelper, and I initially tried to limit the PR to not change that implementation. Also, relativeSD is the user-facing config option, and I wanted to lean into a more strict compatibility check of the config associated with sketches, vs comparing the derived P values (which as you point out could be the same for similar relativeSD values, and result in a less strict compatibility check). This also gives the benefit that the compatibility check's error message reports the user-facing config value associated with the sketch, vs reporting a derived P value.

**Comment by dtenedor** [comment]
Hi @RyanBerti, thanks again for your work on this. There are some high level design questions open on https://github.com/apache/spark/pull/39678. We should probably resolve those before moving forward with the implementation. 🤔

**Comment by RyanBerti** [comment]
@dtenedor ya - I got the feeling that this implementation might be blocked by the larger discussion. My take is that most Spark users probably don't notice/care which HLL code is used by approx_distinct_count (given we've been running the non-Datasketches version since ~2016), unless they're trying to produce/consume approx_distinct_count sketches with different platforms (IE Spark->Presto). That being said, I understand that it makes sense not to introduce new features based on the current HLL implementation if there's an expectation that we'd move to the Datasketches implementation in the short term. Given I didn't find any discussion/design or JIRAs associated with modifying the underling HLL implementation, I hadn't expected it to be a blocker for this implementation. 

I'll be on vacation for the first two weeks of February, but am open to putting in some work on the Datasketches migration after I return late-Feb/March, assuming you'd be open to providing feedback/guidance on that work? My main goal is approx_distinct_count sketch re-aggregation, so I'm hoping we could design implementation/scope the work such that it achieves my goal and leaves the door open for Spark to support more complex types of sketches/aggregations with future work.

**Comment by dtenedor** [comment]
@RyanBerti yes, this is totally valid. To be honest, your PR and email threads sparked a discussion among people who had worked with HLL before in other systems, to try to figure out a good future direction for this feature. You are definitely correct that there were no prior designs or Jiras or anything to this effect yet, we are currently in the process of figuring this out.

I'd be happy to help with the review of the Datasketches-based implementation if/when you have time to help with that. This time frame works well, and once again, your work is appreciated and will be used for a long time in Spark.

It seems for now, maybe the open question is if and how we can generate an implementation based on the existing HyperLogLogPlusPlus helper in such a manner that we don't cause future compatibility problems with the Datasketches-based version later. Brainstorming, one way to do that could be to give these new functions separate names that clearly indicate that a different algorithm is being used, reserving the core "HLL_COUNT.INIT"/"HLL_COUNT.MERGE/"HLL_COUNT.FINALIZE" function names later for the Datasketches-based version. There could still be a risk of users accidentally producing sketches in one format and consuming them with the other, but perhaps the separate function names could help with that. More internal discussion could help figure this out. I can post shortly with an update.

**Comment by dtenedor** [comment]
@RyanBerti it seems that while it could be technically possible to add two implementations, there is a lot of concern about adding complexity for users in this case. For reference, Spark already has some notion amongst users that it is too complex with many knobs and configurations to understand. It seems we're better off taking a bit more time and focusing on the single integration with Datasketches. It is OK for this to take longer as there is a bit more work involved.

**Comment by RyanBerti** [comment]
@dtenedor ok, sounds good. Would you mind creating a new JIRA ticket to track the Datasketches integration, and we can close out https://issues.apache.org/jira/browse/SPARK-16484, along with these PRs, in favor of the new ticket?

**Comment by RyanBerti** [comment]
Closed in favor of https://github.com/apache/spark/pull/40615
